### PR TITLE
MAINT: update pyproject.toml for compatibility with numpy 2

### PR DESCRIPTION
**Pull request by nghi-truyen**
- Add lower bounds to several packages in pyproject for the compatibility with numpy 2.0.
- Remove version bounds in requirement.txt and environment-dev.yml (developers are required to update their pip and conda to ensure that they have the latest package versions).
- Add a reminder in contributor guide (detail development process) for pip and conda update before compiling smash.

### CODE REVIEW

File: doc/source/contributor_guide/development_process_details.rst [hunk 1]
```diff
@@ -19,7 +19,7 @@ System-level dependencies
 `smash` uses compiled code for speed, which means you need compilers and some other system-level
 (i.e, non-Python / non-PyPI) dependencies to build it on your system.
 
-Anaconda (recommanded)
+Anaconda (recommended)
```
Comment (inoelloc) [inline doc/source/contributor_guide/development_process_details.rst:22]: :3

File: pyproject.toml [hunk 2]
```diff
@@ -1,11 +1,7 @@
 [build-system]
 requires = [
     "meson-python",
-
-    # numpy requirement for wheel builds for distribution on PyPI - building
-    # against 2.x yields wheels that are also compatible with numpy 1.x at
-    # runtime.
-    "numpy>=2.0.0",
+    "numpy>=1.24",
```
Comment (inoelloc) [inline pyproject.toml:4]: Lui faut laisser 2.0. On veut absolument build avec numpy>=2.0. (avec le commentaire please)

File: doc/source/contributor_guide/development_process_details.rst [hunk 3]
```diff
@@ -35,6 +35,14 @@ It will create a Conda environment called ``smash-dev`` that can be activated as
 
     conda activate smash-dev
 
+Once ``smash-dev`` is created, it is recomended to update pip and your Conda environment
```
Comment (inoelloc) [inline doc/source/contributor_guide/development_process_details.rst:38]:
```suggestion
Once ``smash-dev`` is created, it is recommended to update pip and your Conda environment
```
C'est dur recommended aha

File: pyproject.toml (hunk 2 above)
Comment (nghi-truyen) [inline pyproject.toml:4]:
```suggestion
    # numpy requirement for wheel builds for distribution on PyPI - building
    # against 2.x yields wheels that are also compatible with numpy 1.x at
    # runtime.
    "numpy>=2.0.0",
```

**Review by inoelloc — APPROVED** [review]

**Pull request merged**